{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a8ee17fa",
   "metadata": {},
   "source": [
    "- Purpose:\n",
    "    - this work uses the comprehensive primary assembly GENCODEv48 gtf\n",
    "    - the GENCODEv48 transcripts fasta file contains transcripts not in the primary assembly\n",
    "    - this can create agreement issues particularly with salmon read counts\n",
    "    - this script writes a new fasta file which filtered to match the gtf "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e60bf4e6",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'os' has no attribute '__version__'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[8]\u001b[39m\u001b[32m, line 10\u001b[39m\n\u001b[32m      8\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpandas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpd\u001b[39;00m\n\u001b[32m      9\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mos\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m10\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[43mos\u001b[49m\u001b[43m.\u001b[49m\u001b[43m__version__\u001b[49m)\n",
      "\u001b[31mAttributeError\u001b[39m: module 'os' has no attribute '__version__'"
     ]
    }
   ],
   "source": [
    "#####################\n",
    "# import statements #\n",
    "#####################\n",
    "from tqdm import tqdm\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import pysam\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "583f635e",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################\n",
    "# User-Defined Variables #\n",
    "##########################\n",
    "# - define all variables below with paths to the required files\n",
    "# - this should be the only cell that requires modification\n",
    "# - the fastagz file should be the transcripts fastagz\n",
    "# - both files should be compressed with bgzip and indexed\n",
    "\n",
    "fastagz_path = '/Volumes/TOSHIBA-EXT/sequencing-datasets/GENCODE/v48/primary-assembly-bgzip/transcript-gtf-match-test/gencode.v48.transcripts.bgzip.fa.gz'\n",
    "gtfgz_path = '/Volumes/TOSHIBA-EXT/sequencing-datasets/GENCODE/v48/primary-assembly-bgzip/transcript-gtf-match-test/gencode.v48.primary_assembly.sorted.bgzip.gtf.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "89eb1657",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000832824.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000832825.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000832826.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000832827.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000832828.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387949</th>\n",
       "      <td>ENST00000797723.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387950</th>\n",
       "      <td>ENST00000819228.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387951</th>\n",
       "      <td>ENST00000819229.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387952</th>\n",
       "      <td>ENST00000751368.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387953</th>\n",
       "      <td>ENST00000839853.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>387954 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id\n",
       "0       ENST00000832824.1\n",
       "1       ENST00000832825.1\n",
       "2       ENST00000832826.1\n",
       "3       ENST00000832827.1\n",
       "4       ENST00000832828.1\n",
       "...                   ...\n",
       "387949  ENST00000797723.1\n",
       "387950  ENST00000819228.1\n",
       "387951  ENST00000819229.1\n",
       "387952  ENST00000751368.1\n",
       "387953  ENST00000839853.1\n",
       "\n",
       "[387954 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# define output directory and transcript fasta name based on user-defined variables\n",
    "out_dir = os.path.split(fastagz_path)[0]\n",
    "fastagz_name = os.path.split(fastagz_path)[1].split('.fa.gz')[0]\n",
    "\n",
    "# build a dataframe of all transcript ids in the reference gtf\n",
    "gtfz = pysam.TabixFile(gtfgz_path)\n",
    "gtfgz_transcript_list = []\n",
    "for gtfgz_line in gtfz.fetch():\n",
    "    gtfgz_rec = gtfgz_line.strip().split('\\t')\n",
    "    feat_type = gtfgz_rec[2].strip()\n",
    "    if feat_type == 'transcript':\n",
    "        gtfgz_transcript_list.append(gtfgz_rec[8].strip().split('transcript_id \"')[1].strip().split('\"')[0].strip())\n",
    "\n",
    "gtfgz_df = pd.DataFrame({\n",
    "    'transcript_id':gtfgz_transcript_list,\n",
    "})\n",
    "\n",
    "# verify that transcript id df seems correct\n",
    "display(gtfgz_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cf59bfe5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000832824.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000832825.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000832826.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000832827.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000832828.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412039</th>\n",
       "      <td>ENST00000797959.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412040</th>\n",
       "      <td>ENST00000819228.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412041</th>\n",
       "      <td>ENST00000819229.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412042</th>\n",
       "      <td>ENST00000751368.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412043</th>\n",
       "      <td>ENST00000839853.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>412044 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id\n",
       "0       ENST00000832824.1\n",
       "1       ENST00000832825.1\n",
       "2       ENST00000832826.1\n",
       "3       ENST00000832827.1\n",
       "4       ENST00000832828.1\n",
       "...                   ...\n",
       "412039  ENST00000797959.1\n",
       "412040  ENST00000819228.1\n",
       "412041  ENST00000819229.1\n",
       "412042  ENST00000751368.1\n",
       "412043  ENST00000839853.1\n",
       "\n",
       "[412044 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create a dataframe that includes all the transcript ids in the transcript fasta\n",
    "fatagz_transcript_list = []\n",
    "with gzip.open(fastagz_path, 'rt') as fastagz:\n",
    "    for fa_rec in SeqIO.parse(fastagz, 'fasta'):\n",
    "        transcript_id = fa_rec.id.strip().split('|')[0].strip()\n",
    "        fatagz_transcript_list.append(transcript_id)\n",
    "\n",
    "fastagz_df = pd.DataFrame({\n",
    "    'transcript_id':fatagz_transcript_list,\n",
    "})\n",
    "\n",
    "# verify that transcript id df seems correct\n",
    "# if the dataframes are different lengths it indicates that the gtf and fasta have different transcript sets\n",
    "display(fastagz_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c14b86ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000832824.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000832825.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000832826.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000832827.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000832828.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387949</th>\n",
       "      <td>ENST00000797723.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387950</th>\n",
       "      <td>ENST00000819228.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387951</th>\n",
       "      <td>ENST00000819229.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387952</th>\n",
       "      <td>ENST00000751368.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387953</th>\n",
       "      <td>ENST00000839853.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>387954 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id\n",
       "0       ENST00000832824.1\n",
       "1       ENST00000832825.1\n",
       "2       ENST00000832826.1\n",
       "3       ENST00000832827.1\n",
       "4       ENST00000832828.1\n",
       "...                   ...\n",
       "387949  ENST00000797723.1\n",
       "387950  ENST00000819228.1\n",
       "387951  ENST00000819229.1\n",
       "387952  ENST00000751368.1\n",
       "387953  ENST00000839853.1\n",
       "\n",
       "[387954 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# merge the gtf and fasta transcript id datasets to create a dataframe containing only the shared transcript ids\n",
    "shared_df = pd.merge(left=gtfgz_df, right=fastagz_df, on=['transcript_id'], how='inner')\n",
    "\n",
    "# verify that transcript id df seems correct\n",
    "# this dataframe should be the same length as the shorter of the two input dfs\n",
    "display(shared_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "869bc6f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "412044it [00:04, 99882.82it/s] "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fasta records for filtered file: 387954\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# iterate through the fasta file and create a list of all fasta records that match the shared df\n",
    "fa_rec_list = []\n",
    "shared_transcript_set = set(shared_df['transcript_id'].values)\n",
    "with gzip.open(fastagz_path, 'rt') as fastagz:\n",
    "    for fa_rec in tqdm(SeqIO.parse(fastagz, 'fasta')):\n",
    "        transcript_id = fa_rec.id.strip().split('|')[0].strip()\n",
    "        transcript_id_set = {transcript_id}\n",
    "        if len(transcript_id_set.intersection(shared_transcript_set)) > 0:\n",
    "            fa_rec_list.append(fa_rec)\n",
    "\n",
    "# the length of the record list should match the shared df\n",
    "print(f'fasta records for filtered file: {len(fa_rec_list)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ef06d78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the list of shared records to a new fasta file\n",
    "# all records in the new fata file will have corresponding annotation in the gtf\n",
    "fasta_path = os.path.join(out_dir, f'{fastagz_name}.gtf-matched.fa')\n",
    "out_fastagz_path = f'{fasta_path.split('.fa')[0]}.gz'\n",
    "\n",
    "SeqIO.write(fa_rec_list, fasta_path, 'fasta')\n",
    "pysam.tabix_compress(filename_in=fasta_path, filename_out=out_fastagz_path)\n",
    "pysam.faidx(out_fastagz_path)\n",
    "os.remove(path=fasta_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioconda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
