{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f9b2ff2f",
   "metadata": {},
   "source": [
    "- Purpose:\n",
    "    - create a transcript id to gene id two-column map with no headers that can be used for salmon\n",
    "    - create a table with trascript id gene id and gene name useful for DEG and other downstream analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e79e58fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#####################\n",
    "# import statements #\n",
    "#####################\n",
    "import pysam\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "import gzip\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2378455",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################\n",
    "# User-Defined Variables #\n",
    "##########################\n",
    "# - define all variables below with paths to the required files\n",
    "# - this should be the only cell that requires modification\n",
    "# - both files should be compressed with bgzip and indexed\n",
    "\n",
    "gtfgz_path = ''\n",
    "transcript_fa_path = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cb799d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set output directory based on gtf path\n",
    "out_dir = os.path.split(gtfgz_path)[0]\n",
    "\n",
    "# read gtf and print the first 20 entries\n",
    "gtfgz = pysam.TabixFile(gtfgz_path)\n",
    "\n",
    "count = 0\n",
    "for gtf_line in gtfgz.fetch():\n",
    "    print(gtf_line)\n",
    "    if (count := count + 1) > 19:\n",
    "        break\n",
    "\n",
    "gtf_feature_count = 0\n",
    "for gtf_line in gtfgz.fetch():\n",
    "    gtf_feature_count += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4691c92f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from the gtf, compile matched lists of gene ids, gene names, and gene types for all gene features\n",
    "gene_list = []\n",
    "gene_name_list = []\n",
    "gene_type_list = []\n",
    "\n",
    "for gtf_line in tqdm(gtfgz.fetch(),total=gtf_feature_count):\n",
    "    gtf_rec = gtf_line.strip().split('\\t')\n",
    "    if gtf_rec[2] == 'gene':\n",
    "        gene_id = gtf_rec[8].strip().split('gene_id \"')[1].strip().split('\"')[0].strip()\n",
    "        gene_name = gtf_rec[8].strip().split('gene_name \"')[1].strip().split('\"')[0].strip()\n",
    "        gene_type = gtf_rec[8].strip().split('gene_type \"')[1].strip().split('\"')[0].strip() if 'gene_type' in gtf_rec[8] else ''\n",
    "\n",
    "        gene_list.append(gene_id)\n",
    "        gene_name_list.append(gene_name)\n",
    "        gene_type_list.append(gene_type)\n",
    "\n",
    "# compile lists into a dataframe\n",
    "gene_map_df = pd.DataFrame({\n",
    "    'gene_id':gene_list,\n",
    "    'gene_name':gene_name_list,\n",
    "    'gene_type':gene_type_list,\n",
    "})\n",
    "display(gene_map_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "613dbc7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from the gtf, compile matched lists of transcript ids, gene ids, and gene names for all transcript features\n",
    "transcript_list = []\n",
    "gene_list = []\n",
    "gene_name_list = []\n",
    "\n",
    "for gtf_line in tqdm(gtfgz.fetch(),total=gtf_feature_count):\n",
    "    gtf_rec = gtf_line.strip().split('\\t')\n",
    "    if gtf_rec[2] == 'transcript':\n",
    "        transcript_list.append(gtf_rec[8].strip().split('transcript_id \"')[1].strip().split('\"')[0].strip())\n",
    "        gene_list.append(gtf_rec[8].strip().split('gene_id \"')[1].strip().split('\"')[0].strip())\n",
    "        gene_name_list.append(gtf_rec[8].strip().split('gene_name \"')[1].strip().split('\"')[0].strip())\n",
    "\n",
    "# compile the transcript ids and gene ids into map_df and write an otput tsv of two columns with no headers that maps transcript_id to gene_id\n",
    "# used as a salmon input\n",
    "map_df = pd.DataFrame({\n",
    "    'transcript_id':transcript_list,\n",
    "    'gene_id':gene_list,\n",
    "})\n",
    "map_tsv_path = os.path.join(out_dir, 'transcript-gene-map.tsv')\n",
    "map_df.to_csv(map_tsv_path, index=False, header=False, sep='\\t')\n",
    "\n",
    "# add gene names as a third column to the df\n",
    "map_df['gene_name'] = gene_name_list\n",
    "display(map_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a9cd68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from the transcripts fasta, compile matched lists of transcript ids, gene ids, and gene names for all gene transcript features\n",
    "transcript_contigs_transcript_list = []\n",
    "transcript_contigs_gene_list = []\n",
    "transcript_contigs_gene_name_list = []\n",
    "\n",
    "with gzip.open(transcript_fa_path, 'rt') as fasta:\n",
    "    for fa_rec in SeqIO.parse(fasta, 'fasta'):\n",
    "        transcript_contigs_rec = fa_rec.id.strip().split('|')\n",
    "        transcript_contigs_transcript_list.append(transcript_contigs_rec[0].strip())\n",
    "        transcript_contigs_gene_list.append(transcript_contigs_rec[1].strip())\n",
    "        transcript_contigs_gene_name_list.append(transcript_contigs_rec[5].strip())\n",
    "\n",
    "# compile into dataframe\n",
    "transcript_contigs_df = pd.DataFrame({\n",
    "    'transcript_id':transcript_contigs_transcript_list,\n",
    "    'gene_id':transcript_contigs_gene_list,\n",
    "    'gene_name':transcript_contigs_gene_name_list,\n",
    "})\n",
    "display(transcript_contigs_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65e70729",
   "metadata": {},
   "outputs": [],
   "source": [
    "# verify that gtf and fasta transcripts match, ideally in the output all three counts will be the same and the unique entries dataframes will be empty\n",
    "print(f'Transcripts from gtf:          \\t{len(map_df)}')\n",
    "print(f'Transcripts from transcript fa:\\t{len(transcript_contigs_df)}')\n",
    "print(f'Transcripts in both:           \\t{len(pd.merge(left=map_df,right=transcript_contigs_df,on=['transcript_id'],how='inner'))}')\n",
    "print(f'Entries unique to transcript fa:\\n')\n",
    "unique_to_fa_df = transcript_contigs_df[~(transcript_contigs_df['transcript_id'].isin(map_df['transcript_id']))]\n",
    "display(unique_to_fa_df)\n",
    "print(f'Entries unique to gtf:\\n')\n",
    "unique_to_gtf_df = map_df[~(map_df['transcript_id'].isin(transcript_contigs_df['transcript_id']))]\n",
    "display(unique_to_gtf_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d44af8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_tsv_path = os.path.join(out_dir, 'transcript-gene-map-w-names.tsv')\n",
    "map_df.to_csv(map_tsv_path, index=False, sep='\\t')  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioconda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
