{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f9b2ff2f",
   "metadata": {},
   "source": [
    "- Purpose:\n",
    "    - create a transcript id to gene id two-column map with no headers that can be used for salmon\n",
    "    - create a table with trascript id gene id and gene name useful for DEG and other downstream analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e79e58fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#####################\n",
    "# import statements #\n",
    "#####################\n",
    "import pysam\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "import gzip\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2378455",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################\n",
    "# User-Defined Variables #\n",
    "##########################\n",
    "# - define all variables below with paths to the required files\n",
    "# - this should be the only cell that requires modification\n",
    "# - both files should be compressed with bgzip and indexed\n",
    "\n",
    "gtfgz_path = ''\n",
    "transcript_fa_path = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9cb799d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\tHAVANA\texon\t11121\t11211\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832824.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-260\"; exon_number 1; exon_id \"ENSE00004248723.1\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\tgene\t11121\t24894\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; level 2; tag \"overlaps_pseudogene\";\n",
      "chr1\tHAVANA\ttranscript\t11121\t14413\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832824.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-260\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\texon\t11125\t11211\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832825.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-261\"; exon_number 1; exon_id \"ENSE00004248721.1\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\ttranscript\t11125\t14405\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832825.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-261\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\texon\t11410\t11671\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832826.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-262\"; exon_number 1; exon_id \"ENSE00004248726.1\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\ttranscript\t11410\t14413\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832826.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-262\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\texon\t11411\t11671\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832827.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-263\"; exon_number 1; exon_id \"ENSE00004248710.1\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\ttranscript\t11411\t14413\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832827.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-263\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\texon\t11426\t11671\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832828.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-264\"; exon_number 1; exon_id \"ENSE00004248702.1\"; level 2; tag \"basic\"; tag \"TAGENE\";\n",
      "chr1\tHAVANA\ttranscript\t11426\t14409\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832828.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-264\"; level 2; tag \"basic\"; tag \"TAGENE\";\n",
      "chr1\tHAVANA\texon\t11770\t11844\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832829.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-265\"; exon_number 1; exon_id \"ENSE00004248719.1\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\ttranscript\t11770\t14416\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832829.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-265\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\texon\t11819\t11844\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832830.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-266\"; exon_number 1; exon_id \"ENSE00004248695.1\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\ttranscript\t11819\t14413\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832830.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-266\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\texon\t11823\t11844\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832837.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-273\"; exon_number 1; exon_id \"ENSE00004248717.1\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\ttranscript\t11823\t14406\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832837.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-273\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\texon\t11824\t11844\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832831.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-267\"; exon_number 1; exon_id \"ENSE00004248709.1\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\texon\t11824\t11844\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832832.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-268\"; exon_number 1; exon_id \"ENSE00004248709.1\"; level 2; tag \"TAGENE\";\n",
      "chr1\tHAVANA\texon\t11824\t11844\t.\t+\t.\tgene_id \"ENSG00000290825.2\"; transcript_id \"ENST00000832833.1\"; gene_type \"lncRNA\"; gene_name \"DDX11L16\"; transcript_type \"lncRNA\"; transcript_name \"DDX11L16-269\"; exon_number 1; exon_id \"ENSE00004248709.1\"; level 2; tag \"TAGENE\";\n"
     ]
    }
   ],
   "source": [
    "# set output directory based on gtf path\n",
    "out_dir = os.path.split(gtfgz_path)[0]\n",
    "\n",
    "# read gtf and print the first 20 entries\n",
    "gtfgz = pysam.TabixFile(gtfgz_path)\n",
    "\n",
    "count = 0\n",
    "for gtf_line in gtfgz.fetch():\n",
    "    print(gtf_line)\n",
    "    if (count := count + 1) > 19:\n",
    "        break\n",
    "\n",
    "gtf_feature_count = 0\n",
    "for gtf_line in gtfgz.fetch():\n",
    "    gtf_feature_count += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4691c92f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 4119244/4119244 [00:03<00:00, 1092779.69it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000223972.6</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000310526.1</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000227232.6</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78889</th>\n",
       "      <td>ENSG00000303867.1</td>\n",
       "      <td>ENSG00000303867</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78890</th>\n",
       "      <td>ENSG00000303902.1</td>\n",
       "      <td>ENSG00000303902</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78891</th>\n",
       "      <td>ENSG00000306528.1</td>\n",
       "      <td>ENSG00000306528</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78892</th>\n",
       "      <td>ENSG00000297844.1</td>\n",
       "      <td>ENSG00000297844</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78893</th>\n",
       "      <td>ENSG00000309258.1</td>\n",
       "      <td>ENSG00000309258</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78894 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_id        gene_name                           gene_type\n",
       "0      ENSG00000290825.2         DDX11L16                              lncRNA\n",
       "1      ENSG00000223972.6          DDX11L1  transcribed_unprocessed_pseudogene\n",
       "2      ENSG00000310526.1           WASH7P                              lncRNA\n",
       "3      ENSG00000227232.6           WASH7P  transcribed_unprocessed_pseudogene\n",
       "4      ENSG00000278267.1        MIR6859-1                               miRNA\n",
       "...                  ...              ...                                 ...\n",
       "78889  ENSG00000303867.1  ENSG00000303867                              lncRNA\n",
       "78890  ENSG00000303902.1  ENSG00000303902                              lncRNA\n",
       "78891  ENSG00000306528.1  ENSG00000306528                              lncRNA\n",
       "78892  ENSG00000297844.1  ENSG00000297844                              lncRNA\n",
       "78893  ENSG00000309258.1  ENSG00000309258                              lncRNA\n",
       "\n",
       "[78894 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# from the gtf, compile matched lists of gene ids, gene names, and gene types for all gene features\n",
    "gene_list = []\n",
    "gene_name_list = []\n",
    "gene_type_list = []\n",
    "\n",
    "for gtf_line in tqdm(gtfgz.fetch(),total=gtf_feature_count):\n",
    "    gtf_rec = gtf_line.strip().split('\\t')\n",
    "    if gtf_rec[2] == 'gene':\n",
    "        gene_id = gtf_rec[8].strip().split('gene_id \"')[1].strip().split('\"')[0].strip()\n",
    "        gene_name = gtf_rec[8].strip().split('gene_name \"')[1].strip().split('\"')[0].strip()\n",
    "        gene_type = gtf_rec[8].strip().split('gene_type \"')[1].strip().split('\"')[0].strip() if 'gene_type' in gtf_rec[8] else ''\n",
    "\n",
    "        gene_list.append(gene_id)\n",
    "        gene_name_list.append(gene_name)\n",
    "        gene_type_list.append(gene_type)\n",
    "\n",
    "# compile lists into a dataframe\n",
    "gene_map_df = pd.DataFrame({\n",
    "    'gene_id':gene_list,\n",
    "    'gene_name':gene_name_list,\n",
    "    'gene_type':gene_type_list,\n",
    "})\n",
    "display(gene_map_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "613dbc7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 4119244/4119244 [00:04<00:00, 912631.82it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000832824.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000832825.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000832826.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000832827.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000832828.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387949</th>\n",
       "      <td>ENST00000797723.1</td>\n",
       "      <td>ENSG00000303867.1</td>\n",
       "      <td>ENSG00000303867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387950</th>\n",
       "      <td>ENST00000819228.1</td>\n",
       "      <td>ENSG00000306528.1</td>\n",
       "      <td>ENSG00000306528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387951</th>\n",
       "      <td>ENST00000819229.1</td>\n",
       "      <td>ENSG00000306528.1</td>\n",
       "      <td>ENSG00000306528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387952</th>\n",
       "      <td>ENST00000751368.1</td>\n",
       "      <td>ENSG00000297844.1</td>\n",
       "      <td>ENSG00000297844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387953</th>\n",
       "      <td>ENST00000839853.1</td>\n",
       "      <td>ENSG00000309258.1</td>\n",
       "      <td>ENSG00000309258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>387954 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id            gene_id        gene_name\n",
       "0       ENST00000832824.1  ENSG00000290825.2         DDX11L16\n",
       "1       ENST00000832825.1  ENSG00000290825.2         DDX11L16\n",
       "2       ENST00000832826.1  ENSG00000290825.2         DDX11L16\n",
       "3       ENST00000832827.1  ENSG00000290825.2         DDX11L16\n",
       "4       ENST00000832828.1  ENSG00000290825.2         DDX11L16\n",
       "...                   ...                ...              ...\n",
       "387949  ENST00000797723.1  ENSG00000303867.1  ENSG00000303867\n",
       "387950  ENST00000819228.1  ENSG00000306528.1  ENSG00000306528\n",
       "387951  ENST00000819229.1  ENSG00000306528.1  ENSG00000306528\n",
       "387952  ENST00000751368.1  ENSG00000297844.1  ENSG00000297844\n",
       "387953  ENST00000839853.1  ENSG00000309258.1  ENSG00000309258\n",
       "\n",
       "[387954 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# from the gtf, compile matched lists of transcript ids, gene ids, and gene names for all transcript features\n",
    "transcript_list = []\n",
    "gene_list = []\n",
    "gene_name_list = []\n",
    "\n",
    "for gtf_line in tqdm(gtfgz.fetch(),total=gtf_feature_count):\n",
    "    gtf_rec = gtf_line.strip().split('\\t')\n",
    "    if gtf_rec[2] == 'transcript':\n",
    "        transcript_list.append(gtf_rec[8].strip().split('transcript_id \"')[1].strip().split('\"')[0].strip())\n",
    "        gene_list.append(gtf_rec[8].strip().split('gene_id \"')[1].strip().split('\"')[0].strip())\n",
    "        gene_name_list.append(gtf_rec[8].strip().split('gene_name \"')[1].strip().split('\"')[0].strip())\n",
    "\n",
    "# compile the transcript ids and gene ids into map_df and write an otput tsv of two columns with no headers that maps transcript_id to gene_id\n",
    "# used as a salmon input\n",
    "map_df = pd.DataFrame({\n",
    "    'transcript_id':transcript_list,\n",
    "    'gene_id':gene_list,\n",
    "})\n",
    "map_tsv_path = os.path.join(out_dir, 'transcript-gene-map.tsv')\n",
    "map_df.to_csv(map_tsv_path, index=False, header=False, sep='\\t')\n",
    "\n",
    "# add gene names as a third column to the df\n",
    "map_df['gene_name'] = gene_name_list\n",
    "display(map_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "94a9cd68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000832824.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000832825.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000832826.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000832827.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000832828.1</td>\n",
       "      <td>ENSG00000290825.2</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387949</th>\n",
       "      <td>ENST00000797959.1</td>\n",
       "      <td>ENSG00000303902.1</td>\n",
       "      <td>ENSG00000303902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387950</th>\n",
       "      <td>ENST00000819228.1</td>\n",
       "      <td>ENSG00000306528.1</td>\n",
       "      <td>ENSG00000306528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387951</th>\n",
       "      <td>ENST00000819229.1</td>\n",
       "      <td>ENSG00000306528.1</td>\n",
       "      <td>ENSG00000306528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387952</th>\n",
       "      <td>ENST00000751368.1</td>\n",
       "      <td>ENSG00000297844.1</td>\n",
       "      <td>ENSG00000297844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387953</th>\n",
       "      <td>ENST00000839853.1</td>\n",
       "      <td>ENSG00000309258.1</td>\n",
       "      <td>ENSG00000309258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>387954 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id            gene_id        gene_name\n",
       "0       ENST00000832824.1  ENSG00000290825.2         DDX11L16\n",
       "1       ENST00000832825.1  ENSG00000290825.2         DDX11L16\n",
       "2       ENST00000832826.1  ENSG00000290825.2         DDX11L16\n",
       "3       ENST00000832827.1  ENSG00000290825.2         DDX11L16\n",
       "4       ENST00000832828.1  ENSG00000290825.2         DDX11L16\n",
       "...                   ...                ...              ...\n",
       "387949  ENST00000797959.1  ENSG00000303902.1  ENSG00000303902\n",
       "387950  ENST00000819228.1  ENSG00000306528.1  ENSG00000306528\n",
       "387951  ENST00000819229.1  ENSG00000306528.1  ENSG00000306528\n",
       "387952  ENST00000751368.1  ENSG00000297844.1  ENSG00000297844\n",
       "387953  ENST00000839853.1  ENSG00000309258.1  ENSG00000309258\n",
       "\n",
       "[387954 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# from the transcripts fasta, compile matched lists of transcript ids, gene ids, and gene names for all gene transcript features\n",
    "transcript_contigs_transcript_list = []\n",
    "transcript_contigs_gene_list = []\n",
    "transcript_contigs_gene_name_list = []\n",
    "\n",
    "with gzip.open(transcript_fa_path, 'rt') as fasta:\n",
    "    for fa_rec in SeqIO.parse(fasta, 'fasta'):\n",
    "        transcript_contigs_rec = fa_rec.id.strip().split('|')\n",
    "        transcript_contigs_transcript_list.append(transcript_contigs_rec[0].strip())\n",
    "        transcript_contigs_gene_list.append(transcript_contigs_rec[1].strip())\n",
    "        transcript_contigs_gene_name_list.append(transcript_contigs_rec[5].strip())\n",
    "\n",
    "# compile into dataframe\n",
    "transcript_contigs_df = pd.DataFrame({\n",
    "    'transcript_id':transcript_contigs_transcript_list,\n",
    "    'gene_id':transcript_contigs_gene_list,\n",
    "    'gene_name':transcript_contigs_gene_name_list,\n",
    "})\n",
    "display(transcript_contigs_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "65e70729",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcripts from gtf:          \t387954\n",
      "Transcripts from transcript fa:\t387954\n",
      "Transcripts in both:           \t387954\n",
      "Entries unique to transcript fa:\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [transcript_id, gene_id, gene_name]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entries unique to gtf:\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [transcript_id, gene_id, gene_name]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# verify that gtf and fasta transcripts match, ideally in the output all three counts will be the same and the unique entries dataframes will be empty\n",
    "print(f'Transcripts from gtf:          \\t{len(map_df)}')\n",
    "print(f'Transcripts from transcript fa:\\t{len(transcript_contigs_df)}')\n",
    "print(f'Transcripts in both:           \\t{len(pd.merge(left=map_df,right=transcript_contigs_df,on=['transcript_id'],how='inner'))}')\n",
    "print(f'Entries unique to transcript fa:\\n')\n",
    "unique_to_fa_df = transcript_contigs_df[~(transcript_contigs_df['transcript_id'].isin(map_df['transcript_id']))]\n",
    "display(unique_to_fa_df)\n",
    "print(f'Entries unique to gtf:\\n')\n",
    "unique_to_gtf_df = map_df[~(map_df['transcript_id'].isin(transcript_contigs_df['transcript_id']))]\n",
    "display(unique_to_gtf_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0d44af8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_tsv_path = os.path.join(out_dir, 'transcript-gene-map-w-names.tsv')\n",
    "map_df.to_csv(map_tsv_path, index=False, sep='\\t')  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioconda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
