{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a8ee17fa",
   "metadata": {},
   "source": [
    "- Purpose:\n",
    "    - this work uses the comprehensive primary assembly GENCODEv48 gtf\n",
    "    - the GENCODEv48 transcripts fasta file contains transcripts not in the primary assembly\n",
    "    - this can create agreement issues particularly with salmon read counts\n",
    "    - this script writes a new fasta file which filtered to match the gtf "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e60bf4e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#####################\n",
    "# import statements #\n",
    "#####################\n",
    "from tqdm import tqdm\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import pysam\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "583f635e",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################\n",
    "# User-Defined Variables #\n",
    "##########################\n",
    "# - define all variables below with paths to the required files\n",
    "# - this should be the only cell that requires modification\n",
    "# - the fastagz file should be the transcripts fastagz\n",
    "# - both files should be compressed with bgzip and indexed\n",
    "\n",
    "fastagz_path = ''\n",
    "gtfgz_path = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89eb1657",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define output directory and transcript fasta name based on user-defined variables\n",
    "out_dir = os.path.split(fastagz_path)[0]\n",
    "fastagz_name = os.path.split(fastagz_path)[1].split('.fa.gz')[0]\n",
    "\n",
    "# build a dataframe of all transcript ids in the reference gtf\n",
    "gtfz = pysam.TabixFile(gtfgz_path)\n",
    "gtfgz_transcript_list = []\n",
    "for gtfgz_line in gtfz.fetch():\n",
    "    gtfgz_rec = gtfgz_line.strip().split('\\t')\n",
    "    feat_type = gtfgz_rec[2].strip()\n",
    "    if feat_type == 'transcript':\n",
    "        gtfgz_transcript_list.append(gtfgz_rec[8].strip().split('transcript_id \"')[1].strip().split('\"')[0].strip())\n",
    "\n",
    "gtfgz_df = pd.DataFrame({\n",
    "    'transcript_id':gtfgz_transcript_list,\n",
    "})\n",
    "\n",
    "# verify that transcript id df seems correct\n",
    "display(gtfgz_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf59bfe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dataframe that includes all the transcript ids in the transcript fasta\n",
    "fatagz_transcript_list = []\n",
    "with gzip.open(fastagz_path, 'rt') as fastagz:\n",
    "    for fa_rec in SeqIO.parse(fastagz, 'fasta'):\n",
    "        transcript_id = fa_rec.id.strip().split('|')[0].strip()\n",
    "        fatagz_transcript_list.append(transcript_id)\n",
    "\n",
    "fastagz_df = pd.DataFrame({\n",
    "    'transcript_id':fatagz_transcript_list,\n",
    "})\n",
    "\n",
    "# verify that transcript id df seems correct\n",
    "# if the dataframes are different lengths it indicates that the gtf and fasta have different transcript sets\n",
    "display(fastagz_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c14b86ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge the gtf and fasta transcript id datasets to create a dataframe containing only the shared transcript ids\n",
    "shared_df = pd.merge(left=gtfgz_df, right=fastagz_df, on=['transcript_id'], how='inner')\n",
    "\n",
    "# verify that transcript id df seems correct\n",
    "# this dataframe should be the same length as the shorter of the two input dfs\n",
    "display(shared_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "869bc6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# iterate through the fasta file and create a list of all fasta records that match the shared df\n",
    "fa_rec_list = []\n",
    "shared_transcript_set = set(shared_df['transcript_id'].values)\n",
    "with gzip.open(fastagz_path, 'rt') as fastagz:\n",
    "    for fa_rec in tqdm(SeqIO.parse(fastagz, 'fasta')):\n",
    "        transcript_id = fa_rec.id.strip().split('|')[0].strip()\n",
    "        transcript_id_set = {transcript_id}\n",
    "        if len(transcript_id_set.intersection(shared_transcript_set)) > 0:\n",
    "            fa_rec_list.append(fa_rec)\n",
    "\n",
    "# the length of the record list should match the shared df\n",
    "print(f'fasta records for filtered file: {len(fa_rec_list)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ef06d78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the list of shared records to a new fasta file\n",
    "# all records in the new fata file will have corresponding annotation in the gtf\n",
    "fasta_path = os.path.join(out_dir, f'{fastagz_name}.gtf-matched.fa')\n",
    "out_fastagz_path = f'{fasta_path.split('.fa')[0]}.gz'\n",
    "\n",
    "SeqIO.write(fa_rec_list, fasta_path, 'fasta')\n",
    "pysam.tabix_compress(filename_in=fasta_path, filename_out=out_fastagz_path)\n",
    "pysam.faidx(out_fastagz_path)\n",
    "os.remove(path=fasta_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioconda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
